py/dml/template.py

```diff
@@ -144,6 +156,88 @@ def traits(self):
             return Set().union(
                 *[sup.traits() for (_, sup) in self.spec.templates])
 
+    @functools.cache
+    def get_potential_method_impl_details(self, method_name):
+        '''Provides details about which among this template and its ancestors
+        may provide the highest-rank implementation of the specified method
+        (when not considering any rank higher than that of the template.)
+
+        Returns a tuple `(final, provides_impl, next_candidates)`:
+        * `final`: True only if instantiating the template makes it so that any
+          implementation of the named method besides the one the template may
+          provide or inherit is impossible. (Note: not "if and only if")
+          If `final` is True, that means either the template provides or
+          inherits a single unoverridable and unconditionally defined
+          implementation, or the template unconditionally provides or inherits
+          a non-method member of the given method name.
+        * `provides_impl`: Signifies that the current template itself specifies
+          an implementation of the method (which may be conditionally provided)
+        * `next_candidates`: A list of all hierarchically unrelated ancestor
+          templates that specify (and thus may provide) a highest-rank
+          implementation of the named method when excluding the current
+          template. If the current template unconditionally provides an
+          implementation, this list will be empty.
+
+        Example return values of `t.get_potential_method_impl_details(m)` and
+        their meaning:
+        * `(True, True, [])`
```

Comment:
Given `functools.cache`, safer to make `next_candidates` a tuple.

Reply:
Nice catch!

Reply:
Done.